source/common/http/http2/codec_impl.cc

```diff
@@ -222,7 +223,11 @@ void ConnectionImpl::ServerStreamImpl::encodeHeaders(const ResponseHeaderMap& he
 
 void ConnectionImpl::StreamImpl::encodeTrailersBase(const HeaderMap& trailers) {
   ASSERT(!local_end_stream_);
-  local_end_stream_ = true;
+
+  const bool skip_encoding_empty_trailers =
+      Runtime::runtimeFeatureEnabled("envoy.reloadable_features.skip_encoding_empty_trailers");
+  local_end_stream_ = !(trailers.empty() && skip_encoding_empty_trailers);
```

Comment:
I'm confused why local_end_stream_ is not always true here like it was before. Logically it's the same, right?

Reply:
If we set it true here, we can't call `encodeData` due to assertion here: https://github.com/envoyproxy/envoy/blob/d7c7e9a79eed80afa56a0b05cf6adf0516750d14/source/common/http/http2/codec_impl.cc#L444

Reply:
I would fix the assertion somehow. I think it's incorrect to change local_end_stream_ to not be true here. (You could somehow not fire the assertion in the empty trailer conversion case by having a common encodeData helper function with a boolean param or something like that, up to you)